### ISSUE TITLE
validate_presence_of(...) with allow_nil for association name

### ISSUE THREAD

**Issue by axlekb** [thoughtbot/shoulda-matchers#683]
Was thinking of the use case to validate the presence of a `belongs_to` record, but not make it required.

An example would be: Document belongs to Signer, but the Signer is not added until after the Document is created.

Here's how I would do the validation:

``` ruby
validates :signer, :presence => true, :allow_nil => true
```

Here's how I'd expect to validate it:

``` ruby
it { is_expected.to validate_presence_of(:signer).allow_nil }
```

but `allow_nil` is not a chainable option for `validate_presence_of`, should it be?

**Comment by mcmire** [comment]
Yup! (`allow_blank` should also be a chainable option too.)

**Comment by axlekb** [comment]
This can be closed too!

**Comment by mcmire** [comment]
:sunglasses:

**Comment by aprescott** [comment]
It looks like #695 (which was going to implement this) was closed. Is this feature implemented? I'm looking to use it!

**Comment by mockdeep** [comment]
Just came across this. Would love to have this matcher. We want to have values be either `nil` or anything that's not an empty string.

**Comment by mcmire** [comment]
I'm going to reopen this for now. If someone could re-create the PR but provide a better use case than combining`validate_presence_of` with `belongs_to`, then that would be a good start.

**Comment by aprescott** [comment]
I've encountered this a few times, for cases where I have an attribute that should either be `nil` or a non-`blank?` string:

```ruby
validates :reason, presence: { allow_nil: true }
```

The work-around has been to manually cover a few cases:

```ruby
it { should allow_value(nil).for(:reason) }
it { should_not allow_value("").for(:reason) }
it { should_not allow_value(" ").for(:reason) }
```

I imagine, like the opening description, that this could instead be something like `validate_presence_of(:reason).allow_nil`. It could test the right cases without having to copy-paste this for any attribute that wants to validate in this way.

**Comment by mockdeep** [comment]
Being that we are receiving user input from client-side, which comes back as empty string rather than `null`, we ended up opting for a setter method instead of the validation:

```ruby
def thing=(thing)
  super(thing.presence)
end
```

**Comment by mcmire** [comment]
Sorry for the super long delay on this. This is now available in 4.1.0.
